implemented camera shake

AlexanderEvans opened the pull request:
I added a camera shake script and game object dolly that can follow a target at an offset for 2d.  I also removed some remnants of an attempt to use EZ Camera Shake.

Commit:
implemented camera shake

I added a camera shake script and game object dolly that can follow a target at an offset for 2d.  I also removed some remnants of an attempt to use EZ Camera Shake.

AlexanderEvans commented:
Oh, I also built a simple custom camera dolly object and hard-locked the cine-machine camera to it on follow and aim.  Hopefully it fixes the built in cine-machine camera derping out with it's trans-poser.  If needed I think I can actually build a full alternative to cine-machine given a few weeks(a handful of days of actual work time).  Hopefully this is good enough though.  The current custom dolly is very much designed for 2d, but I can make a 3d orbital dolly if needed.

Dreaming381 commented on Magnum Prandium/Assets/FollowCam.cs:
```diff
@@ -0,0 +1,22 @@
+    void Update () {
+        transform.position = new Vector3(cameraDollyTarget.transform.position.x, cameraDollyTarget.transform.position.y, cameraDollyTarget.transform.position.z - 10);
+	}
+}
```
This script is unnecessary. This behavior can be created without the derpy movement with Cinemachine. Most likely you have your settings wrong. Make sure all your damping settings are set to 0 for a lag-less tracking.

Dreaming381 commented on Magnum Prandium/Assets/ShakyShakeEarthquakeObjectController.cs:
```diff
@@ -0,0 +1,82 @@
+        transform.localPosition = originalPosition;
+        transform.localRotation = originalRotation;
+    }
+}
```
I'll have to run this code to see if it feels good, but Cinemachine has a really good Camera Shake system you are not using. Try adding an impulse listener module to the vcam and triggering an impulse source.

AlexanderEvans replied:
The Cinemachine tracker/dolly keeps derping out spitting out weird positional transposed values, even setting damping to 0 the camera just jumps to keep up with the player occasionally, sometimes triggered when doing a point turn right after a dash, but not always.  Setting it to hard locked tracking fixes that for some reason, but then necessitates a dolly object.  If we want I can implement a slurp and retargeting solution pretty easily to re-implement cinemachines soft follow feature and the hard borders?  The impulse module only really adds the ability to trigger shake when it gets hit by stuff, something that mine can also do by making a call in a colision script, and add axis specific frequency generation(which if we feel is important I guess I could actually pretty easily add to mine).  It seems to use the same "random" wave every time for a given frequency though, which is quite annoying.  It is nice that it can automatically do distance calculations, but doing vector2d.distance(vec1,vec2) and feeding that into mine as magnitude when it is called isn't terribly intensive(and if we want I could VERY easily write a version that just takes a transform parameter and it's own and gets a distance to automatically scale magnitude with).  I just don't really see why one is terribly better over the other.  One thing that's nice about mine is it can be put on any object, not just cameras, so if we wanted to make a barrel shake or something we could do that, and my version does actually also support setting attack, sustain, and decay values(I just called mine like duration, fadeIn, and fadeOut).  One thing that is nice about the cinemachine version though is if a second shake is triggered before the first ends it seems to behave by shaking additively.  I could add this with a some work on the script if we feel it's essential?  I can also set up events and listeners if we want to make shock waves propogate from points, and in addition to that, I could even add a delay from the focal point so that things don't start shaking "until the wave reaches them" but just scaling a delay factor of the calculated distance value.  If we go the listener and transmitter route, then we no longer need to even pass in a transform to get the distance as the transmitter script can do that internally(actually that's probably how cinemachine's works if I had to guess).  Idk, If the listener worked on everything, it could be cool, but as far as it only working on cameras I just don't see a whole lot of reason to use most of the extra features.  Now if we could shake in world objects relative to a shock waves origin too then this starts to make a lot more sense for an implementation.... I kind of want to code this now.  I might re-write this over the weekend to do just that.

AlexanderEvans replied:
Actually, maybe I will just add camera fade control and inter camera dolly slurps and outright replace cinemachine altogether....

Dreaming381 replied:
You're sacrificing a lot by giving up Cinemachine. First, the smoothing algorithms have special weighted damping functions in place that mimic real-world heavy cameras and tracking systems. They are also designed to eliminate movements that could cause motion sickness. Second, the system is well documented, whereas your system is not, and you are spending a bunch of time replacing a perfectly good system just because you can't solve a strange behavior. And third, you are sacrificing all of your cutscene transition capabilities. That stuff is magic and is especially powerful for this style of game. I'm going to try and reproduce your issues.

Commit:
implemented compounding Shake2D couroutine

...I couldn't sleep so I did it...  The couroutine all seem to excecute in bulk between Update() and LateUpdate() in the excecution order, so I took advantage of that to implement the inverse transformations in the Update() function... Once I figured out the excecution order that wasn't so bad, didn't realize that all the coroutines excecuted in the same place in the loop, that made this SUPER easy upon that realization.

Commit:
naming refactor and folder restructure

Commit:
Implemented shake implode event system

We can now shake the camera, and game objects in the world relative to the starting position of the shockwave for shakable objects(Those with a ShakeyShakeEarthquakeObjectController attatched)

AlexanderEvans commented:
I couldn't sleep, so I fixed sh*t and added stuff to the earthquake shock wave script(Which does work with cinemachine cameras again, thanks Dreaming381).  Not being able to figure it out was bothering keeping me awake.  Please let me know if I have done anything that only a mentally insane chimp with half a brain would do in my sleep deprived state, but I think it is good, working, and stable now, at least as far as I can tell.

Commit:
Added delay to shockwave

Added delay to shockwave based on distance from epicenter.  The shockwave propogates out at 10 unity units per second and becomes significantly weaker as distance increases.

Commit:
Combined custom ShakeyShakeEarthquake system with Cinemachine impulse camera shake system

AlexanderEvans commented:
So, I went ahead and combined my system and the cine machine system, but the cine machine system still isn't really as flexible as I would like.  Although I can set up parameters for a random shake profile and I can control the total magnitude of it programmatically via m_AmplitudeGain in the CinemachineImpulseDefinition class in the CinemachineImpulseSource component, BUT, I can't control the individual axis of the shake at call time, I can only change it in the profile which doesn't result in per-shake granular axis component control at run time in the calls.  Still, I went ahead and integrated as much of the cinemachine impulse system as I could without overriding essentially all their classes related to impulse generation to pass down axis information as calls are made.  At the moment I just kind of fake the component magnitude that the other objects use by doing m_AmplitudeGain  = ((magnitudePos + magnitudeRot)/2.0f)

Dreaming381 commented:
It does seem like aliasing occurs if the same source attempts to play multiple clips. Is that your use case? (If so the correct answer is a pool of impulse sources) Otherwise you are over complicating things. I’d focus on cleaning up your code and removing all the unnecessary code. Right now you have unnecessary singletons which are causing more problems than they solve. Also you are wrapping the Cinemachine settings with your own settings which basically do the same thing. Why? All that does is remove the level designer’s ability to tone down the effect per instance because your code overrides the setting globally. Do you want all of your shake settings to be on a component rather than a scriptableObject? If so, subclass CinemachineImpulseSource and provide it with a runtime SignalSourceAsset that you generate and then can modify whenever you create a new impulse.

AlexanderEvans commented:
There are a few reasons I designed mine this way, but the biggest reason is because this wrapper I wrote allows me to shake other objects aside from the camera when there is a shockwave that triggers a shake.  For instance if I have a bunch of barrel sprites and characters in a room that I want to shake in response to an explosion on screen left in addition to the camera I can, and the way my system works it does it's own distance calculation and distance based magnitude damping so that objects close to the epicenter start/stop shaking first, and they shake harder.  The default cinemachine tools only let me shake cameras.  My scripts also allow more optional control over the individual rotation and  position axes magnitudes, at least on objects that shake directly, cinemachine has the limitation I mentioned earlier unfortunately.  I am also overriding it via script because I want to be able to adjust the magnitude at the call, perhaps depending on how much an enemy has charged an attack.  I want it to be future proof for such a scenario.  Because all of this will be tied to mechanic implementation I want it to be overridden by code by default.  You will notice that I do have a regular call to rockTheBoat() that takes no parameters and uses the component default yet in case we do want the level editors to be able to set the per object shake on things, however we just aren't triggering that method at the moment.  So the functionality for that is still there.  Honestly, I'm still not super sure why it was super important to go through the impulse components rather than just shaking the camera dolly, but I haven't removed any functionality from the cinemachine stuff, I would argue that by allowing the script to shake other objects too before hooking into the cinemachine system it is actually more versatile.  I'm not sure what aliasing you are referring to, but I do intend to call generateImpulse() multiple times before it has completely stopped shaking.  I thought the current effect looked good?  As far as the way I am using singletons I really want to force all the scene data and logic, and game data and logic to go through the respective controllers.  I did this so that when we eventually implement saving and more complex multi-scene composites I will have a strong separation.  I'll admit I'm still not really happy with how the system is working, but I prefer it to having a bunch of scriptable objects that need to be manually attached to things.  I also like the ease of doing a "find all references" on static variables compared to scriptable objects.

Dreaming381 commented:
Wait! This is supposed to be a shockwave system? Not just a camera shake system? I think you were the only one who knew that, because it was not obvious at all that was the intent. And while I understand why you don't like my solutions or the default Cinemachine behavior, your solution is still horrible, designer-unfriendly, and breaks Cinemachine group compositing which is really useful for cutscenes. You aren't future proof at all, but are actually making things fragile. If you were to trigger a short generateImpulse during a long generateImpulse (during an earthquake, a rock falls), your long impulse would be immediately cut off. Also, funneling all your data and logic through singletons is a terrible idea and a really easy way to make your game have weak separation and bad coupling. 

What you really want is a PubSubHub system so that each object can react to a shockwave its own way, and handle receiving multiple shockwaves at once as well as selective masks. DM me on Discord a time between 8-10PM Friday or 2 -10PM Saturday when you are free and we can build this system properly with LiveShare. I know exactly what needs to be done to make this clean, designer friendly, code friendly, durable, extendable, and CPU optimized. Until then, I am writing a rant on the evils of singletons in regards to multi-scene setups. I'll try to have that up before Friday.
